src/rate_limit.rs

```diff
@@ -0,0 +1,239 @@
+#[derive(Copy, Clone)]
+struct PeerRecord {
+    online: bool,
+    remaining_calls: u8,
+}
+
+impl PeerRecord {
+    fn new(online: bool, remaining_calls: u8) -> Self {
+        PeerRecord {
+            online,
+            remaining_calls,
+        }
+    }
+}
+
+pub(crate) trait RateLimiter {
+    fn peer_connected(&mut self, peer_key: PublicKey);
+    fn peer_disconnected(&mut self, peer_key: PublicKey);
+    fn peers(&self) -> Vec<PublicKey>;
+    fn query_peer(&mut self, peer_key: PublicKey) -> bool;
+}
+
+// TokenLimiter keeps a map of currently online peers.
+pub(crate) struct TokenLimiter<C: Clock> {
+    peer_map: HashMap<PublicKey, PeerRecord>,
+    clock: C,
+    call_count: u8,
+    call_frequency: Duration,
+    last_update: Instant,
+}
+
+impl<C: Clock> TokenLimiter<C> {
+    pub(crate) fn new(
+        peers: impl Iterator<Item = PublicKey>,
+        call_count: u8,
+        call_frequency: Duration,
+        clock: C,
+    ) -> Self {
+        let mut peer_map: HashMap<PublicKey, PeerRecord> = HashMap::new();
+        for peer in peers {
+            peer_map.insert(peer, PeerRecord::new(true, call_count));
+        }
+
+        let last_update = clock.now();
+        Self {
+            peer_map,
+            clock,
+            call_count,
+            call_frequency,
+            last_update,
+        }
+    }
+
+    fn needs_update(&self) -> bool {
+        self.clock.now().duration_since(self.last_update) >= self.call_frequency
+    }
+
+    fn update(&mut self) {
+        // We can safely delete offline peers because they would have their call count updated at
+        // this point anyway. We want to delete so that we don't allow an infinitely growing queue.
+        self.peer_map.retain(|_, v| v.online);
+
+        // Refresh allowed call counts per-peer that's left online.
+        for (_, v) in self.peer_map.iter_mut() {
+            v.remaining_calls = self.call_count;
+        }
+
+        self.last_update = self.clock.now();
+    }
+
+    fn hit(&mut self, peer_key: PublicKey) -> bool {
+        match self.peer_map.get_mut(&peer_key) {
+            Some(v) => {
+                if v.remaining_calls == 0 {
+                    return false;
+                }
+
+                v.remaining_calls -= 1;
+                true
+            }
+            None => false,
+        }
+    }
+}
+
+impl<C: Clock> RateLimiter for TokenLimiter<C> {
+    fn peer_connected(&mut self, peer_key: PublicKey) {
+        self.peer_map
+            .entry(peer_key)
+            .and_modify(|e| e.online = true)
+            .or_insert(PeerRecord::new(true, self.call_count));
+    }
+
+    fn peer_disconnected(&mut self, peer_key: PublicKey) {
+        self.peer_map
+            .entry(peer_key)
+            .and_modify(|e| e.online = false);
+    }
+
+    fn peers(&self) -> Vec<PublicKey> {
+        self.peer_map
+            .clone()
+            .into_iter()
+            .filter(|p| p.1.online)
```

Comment:
What's the .1 mean here?

Reply:
When we `filter` we get a tuple of `(pubkey, PeerRecord)`, `p.1` accesses the `PeerRecord` element to get online state :)

Reply:
Yeah unfortunately Rust doesn't support named tuples, would need a struct if we _really_ want naming here, but I think it's a short enough method not to bother having a new struct.